operators/pkg/controller/elasticsearch/driver/default.go

```diff
@@ -361,24 +362,39 @@ func (d *defaultDriver) Reconcile(
 	// passed this point, any pods resource listing should check expectations first
 
 	if !esReachable {
-		// We cannot manipulate ES allocation exclude settings if the ES cluster
-		// cannot be reached, hence we cannot delete pods.
-		// Probably it was just created and is not ready yet.
+		// We cannot manipulate ES allocation settings if the ES cluster cannot be reached,
+		// hence we cannot delete pods. Probably the cluster was just created and is not ready yet.
 		// Let's retry in a while.
-		log.Info("ES external service not ready yet for shard migration reconciliation. Requeuing.")
-
+		log.Info("ES external service not ready yet for any HTTP call. Requeuing.")
 		reconcileState.UpdateElasticsearchPending(resourcesState.CurrentPods.Pods())
-
 		return results.WithResult(defaultRequeue)
 	}
 
+	if len(podsState.RunningReady) >= int(es.Spec.NodeCount()) {
```

Comment:
I'm not 100% confident with this condition.
The problem is the following:

When we replace a node (delete one, create another one), we need to:
1. disable shards allocation before deleting the node: this is to avoid all shards to be suddenly replicated on all other nodes while the node is down
2. re-enable shards allocation once the node is back into the cluster

1. is easy since we know when we'll delete a node.
2. is hard since we don't keep track of which node was "restarted": how do we know that one of the nodes we have in the cluster at the moment actually comes from a rolling upgrade?

Also we want to avoid re-enabling shards allocation if the node is still in the process of being restarted.
The best I could come up with is checking the number of nodes in the cluster. In most cases when doing a rolling upgrade we'll have less nodes than when the cluster is "complete": we must not enable shards allocations at this point. If we have the expected number of nodes, chances are all nodes are currently part of the cluster, there is no node being rolled.

I'm not exactly sure how this could conflict with `maxSurge > 0`, where we could have the expected number of nodes even though one node is being restarted.

Open for better implementation ideas :)

Reply:
One refinement could be:
* empty change set means cluster is stable -> enable shard allocation
* `maxSurge = 0` and `RunningReady==Spec.NodeCount()` that is the reuse/restart only case. We are good to enable shard allocation (btw technically we would have to wait until the node joins the ES cluster I believe)
* `maxSurge > 0` and `maxUnavailable > 0` that is the tricky case 
    * first question why would users have that configured other than by mistake?
      *  the only case I can think of is when you have a mixed topology of emptyDir/tempfs and PVCs
    * we should probably in that case do the grow/shrink i.e. non-reuse changes first and only start with the reuse deletions once we have nothing to add and only pvc reuse deletions left (adjusting performable changes accordingly by adding another pass that enforces that invariant)
    * the problem remains that we can only guess that a reuse is in progress when replica shard allocation is disabled and the user could change the spec at any time to add more nodes in which case we would reanble shard allocation too early causing unnecessary IO. But maybe that is OK? The only alternative I see is resorting to annotations on the Elasticsearch resource or keeping state in memory about ongoing operations (which would be lost on crashes or restarts leading again to unnecessary IO as a trade off)